### PR TITLE
Improvements for associative_scan - Autograd

### PR DESCRIPTION
**Pull request by bohnstingl**
This is part of a series of PRs to improve the functionality of the associatve_scan functionality. This specific PR implements the Autograd for associative_scan. This PR has been derived from https://github.com/pytorch/pytorch/pull/129307.

@ydwu4 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

### CODE REVIEW

File: test/functorch/test_control_flow.py [hunk 1]
```diff
@@ -86,32 +89,46 @@ def _fake_while_loop(cond_fn, body_fn, operands):
...
+            pytree.tree_unflatten(carry, carry_spec),
+            pytree.tree_unflatten(xs, inp_spec),
+        )
+        carry, _ = pytree.tree_flatten(carry)
+        y, _ = pytree.tree_flatten(y)
+        result_flat.append(y)
 
     results = [
-        torch.stack([e[leave_ind] for e in op(result_flat)], dim)
+        torch.concatenate([e[leave_ind] for e in op(result_flat)], dim)
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:125]: Why is it a concatenate? if we associative_scan over (4, 2, 3) over dim=0, each subgraph should work on a slice of (2, 3), and the end results should be of shape (4, 2, 3). Anything wrong with this interface?  After the change, does subgraph takes (1, 2, 3) or result becomes (8, 3).

File: torch/_dynamo/variables/higher_order_ops.py [hunk 2]
```diff
@@ -1044,14 +1044,16 @@ def call_function(
...
     ) -> VariableTracker:
+        from torch._higher_order_ops.associative_scan import first_slice_copy
+
         from .builder import wrap_fx_proxy
 
         args, kwargs = LazyVariableTracker.realize_all((args, kwargs))
 
-        def arg_extractor(combine_fn, xs, dim):
-            return combine_fn, xs, dim
+        def arg_extractor(combine_fn, xs):
```
Comment (ydwu4) [inline torch/_dynamo/variables/higher_order_ops.py:1053]: i feel we should split this diff into 2, first is the the dim change, then the autograd.

File: torch/_higher_order_ops/associative_scan.py [hunk 3]
```diff
@@ -153,18 +220,20 @@ def add(x: torch.Tensor, y: torch.Tensor):
     if reverse:
         leaves = [torch.flip(elem, [dim]) for elem in leaves]
 
-    shape = leaves[0].shape
-    ndim = len(shape)
-    dim = utils.canonicalize_dim(ndim, dim)
+    # Move scan dim to 0 and always perform scan on dim 0
+    orig_scan_dim = dim
+    leaves = [shift_source_dim_to_target_dim(elem, int(dim), 0) for elem in leaves]
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py:225]: we might replace `shift_source_dim_to_target_dim` with torch.movedim(elem, int(dim), 0).

File: torch/_higher_order_ops/associative_scan.py [hunk 4]
```diff
@@ -179,16 +248,21 @@ def add(x: torch.Tensor, y: torch.Tensor):
...
     else:
-        result_flat = associative_scan_op(combine_fn, leaves, dim)
+        result_flat = associative_scan_op(combine_fn, leaves)
 
     if reverse:
-        result_flat = [torch.flip(elem, [dim]) for elem in result_flat]
+        result_flat = [torch.flip(elem, [0]) for elem in result_flat]
+
+    result_flat = [
+        shift_source_dim_to_target_dim(elem, 0, orig_scan_dim) for elem in result_flat
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py:259]: shift_source_dim_to_target_dim -> movedim

File: torch/_higher_order_ops/associative_scan.py [hunk 5]
```diff
@@ -179,16 +248,21 @@ def add(x: torch.Tensor, y: torch.Tensor):
...
+    ]
 
     return pytree.tree_unflatten(result_flat, spec)
 
 
+# TODO: Provide inductor support for generic scan
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py:265]: What's missing in inductor for generic scan. Is it the test failure we talked about?

File: torch/_higher_order_ops/associative_scan.py [hunk 6]
```diff
@@ -341,32 +407,413 @@ def trace_associative_scan(
...
+        ctx._mapped_joint_graph = torch.vmap(joint_graph, 0, 0)
+
+        with torch._C._AutoDispatchBelowAutograd():
+            outs = associative_scan_op(fw_graph, xs)
+            ctx.save_for_backward(*(*xs, *outs))
+
+        return (*outs,)
+
+    @staticmethod
+    def backward(ctx, *flat_grads_unmasked):
```
Comment (ydwu4) [inline torch/_higher_order_ops/associative_scan.py:443]: I trust you on this. 

Didn't look into the details of the backward implementation.  Some general thoughts for better testing this: can we use scan to implement a baseline version first and add much more tests to verify the correctness (e.g. nesting cond, scan and associative scan with autograd, more types of ops inside the body of associative scan(e.g. different kinds of view ops, non-continous inputs and outputs.)

**Review by ydwu4 — CHANGES_REQUESTED** [review]

File: test/functorch/test_control_flow.py (hunk 1 above)
Comment (ydwu4) [reply test/functorch/test_control_flow.py:125]: oh, this one should deleted i feel? _fake_scan is now in _higher_order_ops/scan.py.

**Comment by bohnstingl** [comment]
Closing this PR, as it is split into several smaller PRs: https://github.com/pytorch/pytorch/pull/138858, https://github.com/pytorch/pytorch/pull/139864, https://github.com/pytorch/pytorch/pull/139939
